Establish a simple way to instantiate individual locations and collections

sagarpreet-chadha opened the pull request:
Fixes #38 😄 !
And Fixes #39 🎉 !

sagarpreet-chadha commented:
**Input :** 
`
  locations = [[23.1 , 77.1] , [20.1 , 76.1] , [21.111 , 76.111] , [22.111 , 78.111] , [23.1234 , 76.1234] , [24.123456 , 78.123456] , [25.123456 , 77.123456]] ;`

**zoom_level <= 5              :  7 Markers
        zoom_level between 6 and 7   :  5 Markers 
        zoom_level between 8 and 10  :  3 Markers 
        zoom_level >= 11             :  2 Markers**

_[See Black Markers !  (Blue markers are the one fetched from People's API)]_

### 1) When Zoom level = 5 :  
=> 7 Markers 
<img width="1150" alt="screenshot 2019-02-19 at 1 46 33 pm" src="https://user-images.githubusercontent.com/14952645/52999832-d360b500-344c-11e9-870a-029b67d1f05f.png">

### 2) When Zoom level = 6 :
=> 5 Markers
<img width="1147" alt="screenshot 2019-02-19 at 1 48 12 pm" src="https://user-images.githubusercontent.com/14952645/52999902-09059e00-344d-11e9-911a-09da1622a129.png">
 
### 3) When Zoom level = 8 :
=> 3 Markers
<img width="1149" alt="screenshot 2019-02-19 at 1 49 25 pm" src="https://user-images.githubusercontent.com/14952645/52999974-36524c00-344d-11e9-9c52-f8fdf4a7ca53.png">

sagarpreet-chadha commented:
Hi @jywarren , 
I am adding **_source_url_** as an additional input to LBLD in this PR as well .

sagarpreet-chadha commented:
**Blue Markers** =Data from default People's API 
**Black Markers** = Data from locations array passed to LBLD API (Issue #38 )
**Red Markers** = Data fetched from source_url (i have used `/api/srch/taglocations` as an example) 

<img width="1154" alt="screenshot 2019-02-19 at 2 25 38 pm" src="https://user-images.githubusercontent.com/14952645/53002458-c21aa700-3452-11e9-96fd-266aad5d9cf5.png">

sagarpreet-chadha commented:
@jywarren ...kindly review and see all the screenshots above 😄 . Thanks 🎈 !

jywarren commented:
Hi! I'm going to go through this in more detail, but one thing I wanted to suggest was that we make the gh-pages demo display the same collection of test data on 3 different maps, each of which shows a different zoom level, and each of which displays the correct subset of markers. What do you think?

jywarren commented:
...and this is really exciting, thanks @sagarpreet-chadha !!

jywarren commented:
Just starting reviewing, but noticed `.DS_Store` is in here, and think we need to remove it.

jywarren commented on examples/index.html:
```diff
@@ -55,22 +55,30 @@ <h4><b>By entering co-ordinates</b></h4>
+        zoom_level >= 11             :  2 Markers
+      */
+
+      locations = [[23.1 ,      77.1],
```
Do we need to provide a "precision" value too, optionally? I worry about relying on the stored precision of JavaScript values. Or at least we should document the common "gotchas".

sagarpreet-chadha replied:
Currently calculating precision using : 
`afterDecimal = lat.toString().split(".")[1] ;`
I think we should definitely document these 😄 !

jywarren commented on src/blurredLocationDisplay.js:
```diff
@@ -7,11 +7,35 @@ BlurredLocationDisplay = function BlurredLocationDisplay(options) {
 
   options.Interface = options.Interface || require('./ui/Interface.js');
   options.blurredLocation = options.blurredLocation || {};
+  options.PLpeopleAPI = options.PLpeopleAPI || false ;
```
I guess i'm not sure we should customize PL.org completely into here - perhaps better to pass in a custom parser for PL.org API in the example and docs, but not build it into the general-purpose library? What do you think?

jywarren commented on src/blurredLocationDisplay.js:
```diff
@@ -7,11 +7,35 @@ BlurredLocationDisplay = function BlurredLocationDisplay(options) {
   options.blurredLocation = options.blurredLocation || {};
+  options.PLpeopleAPI = options.PLpeopleAPI || false ;
+  options.locations = options.locations || [] ;
+  options.source_url = options.source_url || "" ;
```
We should clearly document what happens if you pass in locations AND pass a source URL.

jywarren commented on src/blurredLocationDisplay.js:
```diff
@@ -7,11 +7,35 @@ BlurredLocationDisplay = function BlurredLocationDisplay(options) {
   InterfaceOptions.blurredLocation = options.blurredLocation;
   var Interface = options.Interface(InterfaceOptions);
 
+  L.Icon.BlackIcon = L.Icon.extend({
```
Should we try to document what we use each color for with a comment in the source?

jywarren commented on src/blurredLocationDisplay.js:
```diff
@@ -72,83 +96,160 @@ BlurredLocationDisplay = function BlurredLocationDisplay(options) {
+  }
+
+  function fetchPLpeopleAPI() {
     map.on('zoomend' , function () {
```
If we run all remote data sources through the same standard handlers, these would become generic.

Also, shouldn't we just be:

1. listening for map interactions
2 triggering fetch from any `source_url`
3. waiting for response
4. storing all, but filtering to decide which are displayed?

or something like that?

sagarpreet-chadha replied:
Yes @jywarren ...will make code the generic 👍  .

We are doing that only : 
1.)    listening for map interactions
https://github.com/sagarpreet-chadha/leaflet-blurred-location-display/blob/ae7806f6089de011473b7ac0998a6f84d8d4ddb9/src/blurredLocationDisplay.js#L205

2.) triggering fetch from any `source_url` and waiting for response
https://github.com/sagarpreet-chadha/leaflet-blurred-location-display/blob/ae7806f6089de011473b7ac0998a6f84d8d4ddb9/src/blurredLocationDisplay.js#L177

3.)  storing all, but filtering to decide which are displayed?

https://github.com/sagarpreet-chadha/leaflet-blurred-location-display/blob/ae7806f6089de011473b7ac0998a6f84d8d4ddb9/src/blurredLocationDisplay.js#L189

AND

https://github.com/sagarpreet-chadha/leaflet-blurred-location-display/blob/ae7806f6089de011473b7ac0998a6f84d8d4ddb9/src/blurredLocationDisplay.js#L66

What do you think ?

jywarren replied:
Awesome I assumed you did this but I think the generic version of the code will make it easier to see this. Let's do it and review again! Thanks Sagarpreet!

sagarpreet-chadha commented:
To Do : 

- [ ]  Proper Documentation .

- [ ]  Make code generic .

- [ ] gh-pages to show 3 maps at different zoom levels .

- [ ] Pass a parser function to LBLD API for `source_url` , else use default parser function .

sagarpreet-chadha commented:
@jywarren , 
Made code more generic 👍 
Removed  default /people API 👍 
Added option to pass JSONparser 👍 

Working on documentation now 😄 . Kindly review again !

sagarpreet-chadha commented:
Demo : https://publiclab.github.io/leaflet-blurred-location-display/examples/example.html

jywarren commented:
This is really nice. Shall we open an issue to plan out having an example of zooming out to show that high precision markers can be made to disappear then?

jywarren commented:
Can we also open a new issue to color code the markers and coordinate text so that people can distinguish the different precisions visually?

And one to default to geoJson for Json parsing?

This is really fantastic Sagarpreet. Good work!
